src/services/RebalanceHLPv2Service.sol

```diff
@@ -0,0 +1,179 @@
+// SPDX-License-Identifier: BUSL-1.1
+// This code is made available under the terms and conditions of the Business Source License 1.1 (BUSL-1.1).
+// The act of publishing this code is driven by the aim to promote transparency and facilitate its utilization for educational purposes.
+
+pragma solidity 0.8.18;
+
+// libs
+import { OwnableUpgradeable } from "@openzeppelin-upgradeable/contracts/access/OwnableUpgradeable.sol";
+import { SafeERC20Upgradeable } from "@openzeppelin-upgradeable/contracts/token/ERC20/utils/SafeERC20Upgradeable.sol";
+import { IERC20Upgradeable } from "@openzeppelin-upgradeable/contracts/token/ERC20/IERC20Upgradeable.sol";
+
+// interfaces
+import { IGmxExchangeRouter } from "@hmx/interfaces/gmx-v2/IGmxExchangeRouter.sol";
+import { IDepositCallbackReceiver, EventUtils, Deposit } from "@hmx/interfaces/gmx-v2/IDepositCallbackReceiver.sol";
+import { IVaultStorage } from "@hmx/storages/interfaces/IVaultStorage.sol";
+import { IConfigStorage } from "@hmx/storages/interfaces/IConfigStorage.sol";
+import { IRebalanceHLPv2Service } from "@hmx/services/interfaces/IRebalanceHLPv2Service.sol";
+
+contract RebalanceHLPv2Service is OwnableUpgradeable, IDepositCallbackReceiver, IRebalanceHLPv2Service {
+  using SafeERC20Upgradeable for IERC20Upgradeable;
+
+  IVaultStorage public vaultStorage;
+  IConfigStorage public configStorage;
+  IGmxExchangeRouter public exchangeRouter;
+  IERC20Upgradeable public weth;
+  address public depositVault;
+  address public depositHandler;
+  uint16 public minHLPValueLossBPS;
+
+  mapping(bytes32 gmxOrderKey => DepositParams depositParam) depositHistory;
+
+  modifier onlyWhitelisted() {
+    configStorage.validateServiceExecutor(address(this), msg.sender);
+    _;
+  }
+
+  modifier onlyGmxDepositHandler() {
+    if (msg.sender != depositHandler) revert IRebalanceHLPv2Service_Unauthorized();
+    _;
+  }
+
+  event LogDepositCreated(bytes32 gmxOrderKey, DepositParams depositParam);
+  event LogSetMinHLPValueLossBPS(uint16 oldValue, uint16 newValue);
+
+  function initialize(
+    IERC20Upgradeable _weth,
+    address _vaultStorage,
+    address _configStorage,
+    address _exchangeRouter,
+    address _depositVault,
+    address _depositHandler,
+    uint16 _minHLPValueLossBPS
+  ) external initializer {
+    OwnableUpgradeable.__Ownable_init();
+
+    weth = _weth;
+    vaultStorage = IVaultStorage(_vaultStorage);
+    configStorage = IConfigStorage(_configStorage);
+    exchangeRouter = IGmxExchangeRouter(_exchangeRouter);
+    depositVault = _depositVault;
+    depositHandler = _depositHandler;
+    minHLPValueLossBPS = _minHLPValueLossBPS;
+  }
+
+  function executeDeposits(
+    DepositParams[] calldata _depositParams,
+    uint256 _executionFee
```

Comment:
Should `executionFee` be inside each `depositParam`? Because each deposit order should have its own execution fee.